rust/agama-cli/src/server.rs

```diff
@@ -0,0 +1,195 @@
+use clap::Subcommand;
+use reqwest::header::{HeaderMap, HeaderValue, CONTENT_TYPE};
+use std::fs::File;
+use std::io;
+use std::io::{BufRead, BufReader};
+use std::path::PathBuf;
+
+const DEFAULT_JWT_FILE: &str = "/tmp/agama-jwt";
+const DEFAULT_AUTH_URL: &str = "http://localhost:3000/authenticate";
+
+#[derive(Subcommand, Debug)]
+pub enum ServerCommands {
+    /// Login with defined server. Result is JWT stored and used in all subsequent commands
+    Login {
+        #[clap(long, short = 'p')]
+        password: Option<String>,
+        #[clap(long, short = 'f')]
+        file: Option<PathBuf>,
+    },
+    /// Release currently stored JWT
+    Logout,
+}
+
+/// Main entry point called from agama CLI main loop
+pub async fn run(subcommand: ServerCommands) -> anyhow::Result<()> {
+    match subcommand {
+        ServerCommands::Login { password, file } => {
+            // actions to do:
+            // 1) somehow obtain credentials (interactive, commandline, from a file)
+            // credentials are handled in this way (in descending priority)
+            // "command line options" -> "read from file" -> "ask to the user"
+            // 2) pass credentials to the web server
+            // 3) receive JWT
+            // 4) store the JWT in a well known location
+
+            let options = LoginOptions {
```

Comment:
Clap can do this for you:

```rust
#[derive(Debug, clap::Args)]
pub enum ServerCommands {
  Login(LoginOptions)
  Logout
}

struct LoginOptions {
  #[clap(long, short = 'p')]
  password: Option<String>,
  #[clap(long, short = 'f')]
  file: Option<String>
}
```

np: and if you follow that approach, you could rename it to `LoginArgs` :-)

Reply:
in the end i managed to get it working ;-)